src/main/drivers/camera_control.c

```diff
@@ -0,0 +1,229 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <platform.h>
+
+#ifdef CAMERA_CONTROL
+
+#include "camera_control.h"
+#include "config/parameter_group_ids.h"
+#include "io.h"
+#include "pwm_output.h"
+#include "system.h"
+#include "math.h"
+#include "nvic.h"
+
+#if defined(STM32F40_41xxx)
+#define CAMERA_CONTROL_TIMER_MHZ   84
+#elif defined(STM32F7)
+#define CAMERA_CONTROL_TIMER_MHZ   108
+#else
+#define CAMERA_CONTROL_TIMER_MHZ   72
+#endif
+
+#define CAMERA_CONTROL_PWM_RESOLUTION   128
+#define CAMERA_CONTROL_SOFT_PWM_RESOLUTION 448
+
+#ifdef CURRENT_TARGET_CPU_VOLTAGE
+#define ADC_VOLTAGE CURRENT_TARGET_CPU_VOLTAGE
+#else
+#define ADC_VOLTAGE 3.3f
+#endif
+
+#if !defined(STM32F411xE) && !defined(STM32F7) && !defined(SITL)
+#define CAMERA_CONTROL_SOFTWARE_PWM_AVAILABLE
+#include "build/atomic.h"
+#endif
+
+#if !defined(SITL)
+#define CAMERA_CONTROL_HARDWARE_PWM_AVAILABLE
+#include "timer.h"
+#endif
+
+
+PG_REGISTER_WITH_RESET_TEMPLATE(cameraControlConfig_t, cameraControlConfig, PG_CAMERA_CONTROL_CONFIG, 0);
+
+PG_RESET_TEMPLATE(cameraControlConfig_t, cameraControlConfig,
+    .mode = CAMERA_CONTROL_MODE_HARDWARE_PWM,
+    .refVoltage = 330,
+    .keyDelayMs = 150,
+    .ioTag = IO_TAG_NONE
+);
+
+static pwmOutputPort_t cameraControlPwm;
+// @todo protect against overflow
+static uint32_t endTimeMillis;
+
+#ifdef CAMERA_CONTROL_SOFTWARE_PWM_AVAILABLE
+void TIM6_DAC_IRQHandler()
+{
+    IOHi(cameraControlPwm.io);
+
+    TIM6->SR = 0;
+}
+
+void TIM7_IRQHandler()
+{
+    IOLo(cameraControlPwm.io);
+
+    TIM7->SR = 0;
+}
+#endif
+
+void cameraControlInit()
+{
+    if (cameraControlConfig()->ioTag == IO_TAG_NONE)
+        return;
+
+    cameraControlPwm.io = IOGetByTag(cameraControlConfig()->ioTag);
+    IOInit(cameraControlPwm.io, OWNER_CAMERA_CONTROL, 0);
+
+    if (CAMERA_CONTROL_MODE_HARDWARE_PWM == cameraControlConfig()->mode) {
+#ifdef CAMERA_CONTROL_HARDWARE_PWM_AVAILABLE
+        const timerHardware_t *timerHardware = timerGetByTag(cameraControlConfig()->ioTag, TIM_USE_ANY);
+
+        if (!timerHardware) {
+            return;
+        }
+
+        #ifdef USE_HAL_DRIVER
+        IOConfigGPIOAF(cameraControlPwm.io, IOCFG_AF_PP, timerHardware->alternateFunction);
+        #else
+        IOConfigGPIO(cameraControlPwm.io, IOCFG_AF_PP);
+        #endif
+
+        pwmOutConfig(&cameraControlPwm, timerHardware, CAMERA_CONTROL_TIMER_MHZ, CAMERA_CONTROL_PWM_RESOLUTION, 0, 0);
```

Comment:
This will prevent sharing of timer with other functions, possibly with unexpected consequences (PWM duty and frequency will change a lot)